sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/GraphExecution.scala

```diff
@@ -0,0 +1,285 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pipelines.graph
+import java.util.concurrent.{ConcurrentHashMap, TimeoutException}
+
+import scala.concurrent.ExecutionContext
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.pipelines.logging.StreamListener
+import org.apache.spark.sql.streaming.Trigger
+
+abstract class GraphExecution(
+    val graphForExecution: DataflowGraph,
+    env: PipelineUpdateContext
+) extends Logging {
+
+  /** The [[Trigger]] configuration for a streaming flow. */
+  def streamTrigger(flow: Flow): Trigger
+
+  protected val pipelineConf: PipelineConf = env.pipelineConf
+
+  /** Maps flowName to count of consecutive failures. Used to manage flow retries */
+  private val flowToNumConsecutiveFailure = new ConcurrentHashMap[TableIdentifier, Int].asScala
+
+  /** Maps flow identifier to count of successful runs. Used to populate batch id. */
+  private val flowToNumSuccess = new ConcurrentHashMap[TableIdentifier, Long].asScala
+
+  /**
+   * [[FlowExecution]]s currently being executed and tracked by this FlowExecution.
+   */
+  val physicalFlows = new collection.concurrent.TrieMap[TableIdentifier, FlowExecution]
+
+  /** Increments flow execution retry count for `flow`. */
+  private def incrementFlowToNumConsecutiveFailure(flowIdentifier: TableIdentifier): Unit = {
+    flowToNumConsecutiveFailure.put(flowIdentifier, flowToNumConsecutiveFailure(flowIdentifier) + 1)
+  }
+
+  /** Protected so tests can override. */
+  private val flowPlanner = new FlowPlanner(
+    graph = graphForExecution,
+    updateContext = env,
+    triggerFor = streamTrigger
+  )
+
+  // Listener to process events and metrics.
+  private val streamListener = new StreamListener(env, graphForExecution)
+
+  /**
+   * Plans the logical [[ResolvedFlow]] into a [[FlowExecution]] and then starts executing it.
+   * Implementation note: Thread safe
+   *
+   * @return None if the flow planner decided that there is no actual update required here.
+   *         Otherwise returns the corresponding physical flow.
+   */
+  def planAndStartFlow(flow: ResolvedFlow): Option[FlowExecution] = {
```

Comment:
this is the first step of an individual flow execution, we plan the resolved logical dataflow into a concrete FlowExecution plan and start the flow execution async. 

upon succeed, we log an event log and move on to the next flow.
upon failure, we log an event log and decided whether it is retryable. If we gonna retry, we push it back to the queue for retry. If not, we mark the flow as failed and mark all the downstream flows as skipped. (other irrelevant flows are not impacted)